### PR TITLE
Add assertionresult and rename

### PR DESCRIPTION
**Pull request by keithamus**
This adds the `AssertionResult` object, plus some helpers.

Also renamed `src/index.ts` to `mod.ts` which is a Deno convention which is more relevant in Deno than `index.js` is to node, especially as Node consumes the `dist/` files.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "allowSyntheticDefaultImports": true,
     "outDir": "dist"
   },
-  "include": [
-    "src"
+  "files": [
+    "mod"
```
Comment (koddsson) [inline tsconfig.json:14]: Should this be:

```suggestion
    "./mod.ts"
```

??

File: mod.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+const startStackFrames = new WeakMap()
+
+interface Result {
+  name: 'AssertionError' | 'AssertionResult'
+  ok: boolean
+  toJSON(...args: unknown[]): Record<string, unknown>
+}
+
+export class AssertionError<T> extends Error implements Result {
+  [key: string]: unknown
```
Comment (koddsson) [inline mod.ts:11]: Is it possible to move this into the `Result` interface?

File: mod.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  toJSON(): Record<string, unknown> {
+    return {
+      ...this,
+      ok: this.ok,
+    }
+  }
+}
+
+const x = new AssertionError<{has: true}>('hi', {has:true})
+x.has === true
```
Comment (koddsson) [inline mod.ts:77]: Remove this?

```suggestion
```

File: mod.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+  get ok() {
+    return true
+  }
+
+  constructor(props?: T, ssf?: Function) {
+    for (const key in props)
+      if (!(key in this))
+        // @ts-ignore
+        this[key] = props[key];
```
Comment (koddsson) [inline mod.ts:65]: I find it hard to read by just indentation. Could we wrap these in curlies?

```suggestion
    for (const key in props) {
      if (!(key in this)) {
        // @ts-ignore
        this[key] = props[key];
      }
    }
```

File: mod.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    return false
+  }
+
+  constructor(public message = 'Unspecified AssertionError', props?: T, ssf?: Function) {
+    super(message)
+    if (canElideFrames && ssf) startStackFrames.set(this, ssf)
+    for (const key in props)
+      if (!(key in this))
+        // @ts-ignore
+        this[key] = props[key];
```
Comment (koddsson) [inline mod.ts:27]: I find it hard to read by just indentation. Could we wrap these in curlies?

```suggestion
    for (const key in props) {
      if (!(key in this)) {
        // @ts-ignore
        this[key] = props[key];
      }
    }
```

File: mod.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    }
+  }
+}
+
+const x = new AssertionError<{has: true}>('hi', {has:true})
+x.has === true
+
+export const ok = <T extends {}>(val: T) => new AssertionResult<T>(val)
+export const error = <E extends {}>(message: string, val: E) => new AssertionError<E>(message, val)
+export default { ok, error }
```
Comment (koddsson) [inline mod.ts:81]: This is interesting. What's the benefit of doing this as opposed to importing `AssertionError` and `AssertionResult`?

**Review by koddsson — APPROVED** [review]

File: mod.ts (hunk 6 above)
Comment (keithamus) [reply mod.ts:82]: It’s just a shorthand. Fine to remove it but thought it might be nice to have.

File: mod.ts (hunk 2 above)
Comment (keithamus) [reply mod.ts:11]: Moving this to the interface causes type errors because then the class doesn't implement this type.

**Pull request merged**